### PR TITLE
core, network: Add ability to proxy socket packets

### PR DESCRIPTION
**Pull request by FearlessTobi**
This is the second part of a series of Pull Requests, aiming to bring multiplayer functionality to yuzu.
It adds the ability to proxy packets coming from our BSD socket implementation and therefore makes it possible to 
play "LAN-supported" games over the internet, without the need for external programs like Hamachi.
For routing it introduces the concept of "fake" ip-addresses. It is needed in order not to leak the users local ip-address and to make multi-instance multiplayer on the same device possible. 

The Pull Request also adds the dedicated-room build-target to make it easy for users to host their own server on either a local machine or a VPS.

Thanks a lot, especially to @german77 for helping me out a lot, @lat9nq for making the code compatible with Linux and everyone who helped me test the feature.

Please note that an LDN implementation is **not** part of this Pull Request.

### CODE REVIEW

File: src/core/internal_network/sockets.h [hunk 1]
```diff
@@ -14,20 +14,92 @@
...
+    virtual Errno SetRcvTimeo(u32 value) = 0;
+
+    virtual Errno SetNonBlock(bool enable) = 0;
+
+    virtual bool IsOpened() const = 0;
+
+    virtual void HandleProxyPacket(const ProxyPacket& packet) = 0;
+
+#if defined(_WIN32)
+    SOCKET fd = INVALID_SOCKET;
```
Comment (liamwhite) [inline src/core/internal_network/sockets.h:93]: Given that you just defined SOCKET and INVALID_SOCKET above, is this really necessary to conditionally compile?

File: src/dedicated_room/yuzu-room.cpp [hunk 2]
```diff
@@ -0,0 +1,375 @@
+// SPDX-FileCopyrightText: Copyright 2017 Citra Emulator Project
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:1]: yuzu_room.cpp, yuzu_room.rc

File: src/dedicated_room/yuzu-room.cpp [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+/// The magic text at the beginning of a yuzu-room ban list file.
+static constexpr char BanListMagic[] = "YuzuRoom-BanList-1";
+
+static constexpr char token_delimiter{':'};
+
+static std::string UsernameFromDisplayToken(const std::string& display_token) {
+    std::size_t outlen;
+
+    std::array<unsigned char, 512> output;
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:79]: default initialize

File: src/dedicated_room/yuzu-room.cpp [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+                          reinterpret_cast<const unsigned char*>(display_token.c_str()),
+                          display_token.length());
+    std::string decoded_display_token(reinterpret_cast<char*>(&output), outlen);
+    return decoded_display_token.substr(0, decoded_display_token.find(token_delimiter));
+}
+
+static std::string TokenFromDisplayToken(const std::string& display_token) {
+    std::size_t outlen;
+
+    std::array<unsigned char, 512> output;
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:90]: default initialize

File: src/dedicated_room/yuzu-room.cpp [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+                 "--web-api-url       yuzu Web API url\n"
+                 "--ban-list-file     The file for storing the room ban list\n"
+                 "--log-file          The file for storing the room log\n"
+                 "--enable-yuzu-mods Allow yuzu Community Moderators to moderate on your room\n"
+                 "-h, --help          Display this help and exit\n"
+                 "-v, --version       Output version information and exit\n";
+}
+
+static void PrintVersion() {
+    std::cout << "yuzu dedicated room " << Common::g_scm_branch << " " << Common::g_scm_desc
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:67]: fmt::print

File: src/dedicated_room/yuzu-room.cpp [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+                          display_token.length());
+    std::string decoded_display_token(reinterpret_cast<char*>(&output), outlen);
+    return decoded_display_token.substr(decoded_display_token.find(token_delimiter) + 1);
+}
+
+static Network::Room::BanList LoadBanList(const std::string& path) {
+    std::ifstream file;
+    Common::FS::OpenFileStream(file, path, std::ios_base::in);
+    if (!file || file.eof()) {
+        std::cout << "Could not open ban list!\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:102]: LOG_ERROR

File: src/dedicated_room/yuzu-room.cpp [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+        return {};
+    }
+    std::string magic;
+    std::getline(file, magic);
+    if (magic != BanListMagic) {
+        std::cout << "Ban list is not valid!\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:108]: LOG_ERROR

File: src/dedicated_room/yuzu-room.cpp [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+    }
+
+    return {username_ban_list, ip_ban_list};
+}
+
+static void SaveBanList(const Network::Room::BanList& ban_list, const std::string& path) {
+    std::ofstream file;
+    Common::FS::OpenFileStream(file, path, std::ios_base::out);
+    if (!file) {
+        std::cout << "Could not save ban list!\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:140]: LOG_ERROR

File: src/dedicated_room/yuzu-room.cpp [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+        file << username << "\n";
+    }
+    file << "\n";
+
+    // IP ban list
+    for (const auto& ip : ban_list.second) {
+        file << ip << "\n";
+    }
+
+    file.flush();
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:157]: It should automatically flush when going out of scope

File: src/dedicated_room/yuzu-room.cpp [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+        std::cout << "preferred game is empty!\n\n";
+        PrintHelp(argv[0]);
+        return -1;
+    }
+    if (preferred_game_id == 0) {
+        std::cout << "preferred-game-id not set!\nThis should get set to allow users to find your "
+                     "room.\nSet with --preferred-game-id id\n\n";
+    }
+    if (max_members > Network::MaxConcurrentConnections || max_members < 2) {
+        std::cout << "max_members needs to be in the range 2 - "
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:273]: LOG_ERROR

File: src/dedicated_room/yuzu-room.cpp [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+                  << Network::MaxConcurrentConnections << "!\n\n";
+        PrintHelp(argv[0]);
+        return -1;
+    }
+    if (port > 65535) {
+        std::cout << "port needs to be in the range 0 - 65535!\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:279]: LOG_ERROR

File: src/dedicated_room/yuzu-room.cpp [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+        PrintHelp(argv[0]);
+        return -1;
+    }
+    if (ban_list_file.empty()) {
+        std::cout << "Ban list file not set!\nThis should get set to load and save room ban "
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:284]: LOG_ERROR

File: src/dedicated_room/yuzu-room.cpp [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+                     "list.\nSet with --ban-list-file <file>\n\n";
+    }
+    bool announce = true;
+    if (token.empty() && announce) {
+        announce = false;
+        std::cout << "token is empty: Hosting a private room\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:290]: LOG_INFO

File: src/dedicated_room/yuzu-room.cpp [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+    }
+    if (web_api_url.empty() && announce) {
+        announce = false;
+        std::cout << "endpoint url is empty: Hosting a private room\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:294]: LOG_INFO

File: src/dedicated_room/yuzu-room.cpp [hunk 15]
```diff
@@ -0,0 +1,375 @@
...
+    }
+    if (announce) {
+        if (username.empty()) {
+            std::cout << "Hosting a public room\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:298]: LOG_INFO

File: src/dedicated_room/yuzu-room.cpp [hunk 16]
```diff
@@ -0,0 +1,375 @@
...
+            Settings::values.web_api_url = web_api_url;
+            Settings::values.yuzu_username = UsernameFromDisplayToken(token);
+            username = Settings::values.yuzu_username.GetValue();
+            Settings::values.yuzu_token = TokenFromDisplayToken(token);
+        } else {
+            std::cout << "Hosting a public room\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:304]: LOG_INFO

File: src/dedicated_room/yuzu-room.cpp [hunk 17]
```diff
@@ -0,0 +1,375 @@
...
+            Settings::values.web_api_url = web_api_url;
+            Settings::values.yuzu_username = username;
+            Settings::values.yuzu_token = token;
+        }
+    }
+    if (!announce && enable_yuzu_mods) {
+        enable_yuzu_mods = false;
+        std::cout << "Can not enable yuzu Moderators for private rooms\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:312]: LOG_WARNING

File: src/dedicated_room/yuzu-room.cpp [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+        ban_list = LoadBanList(ban_list_file);
+    }
+
+    std::unique_ptr<Network::VerifyUser::Backend> verify_backend;
+    if (announce) {
+#ifdef ENABLE_WEB_SERVICE
+        verify_backend =
+            std::make_unique<WebService::VerifyUserJWT>(Settings::values.web_api_url.GetValue());
+#else
+        std::cout
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:329]: LOG_INFO

File: src/dedicated_room/yuzu-room.cpp [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+    Core::System system{};
+    auto& network = system.GetRoomNetwork();
+    network.Init();
+    if (std::shared_ptr<Network::Room> room = network.GetRoom().lock()) {
+        AnnounceMultiplayerRoom::GameInfo preferred_game_info{.name = preferred_game,
+                                                              .id = preferred_game_id};
+        if (!room->Create(room_name, room_description, "", port, password, max_members, username,
+                          preferred_game_info, std::move(verify_backend), ban_list,
+                          enable_yuzu_mods)) {
+            std::cout << "Failed to create room: \n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:346]: LOG_ERROR

File: src/dedicated_room/yuzu-room.cpp [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+            return -1;
+        }
+        std::cout << "Room is open. Close with Q+Enter...\n\n";
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:349]: LOG_INFO

File: src/dedicated_room/yuzu-room.cpp [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+        if (announce) {
+            announce_session->Start();
+        }
+        while (room->GetState() == Network::Room::State::Open) {
+            std::string in;
+            std::cin >> in;
+            if (in.size() > 0) {
+                break;
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (liamwhite) [inline src/dedicated_room/yuzu-room.cpp:360]: I get that this is simpler than having a separate thread to wake you up when the room state changes or when new input is available from the console, but this is unpleasant.

File: src/core/hle/service/nifm/nifm.cpp [hunk 22]
```diff
@@ -271,213 +270,227 @@ class INetworkProfile final : public ServiceFramework<INetworkProfile> {
...
-        ASSERT_MSG(ctx.GetReadBufferSize() == 0x17c,
-                   "SfNetworkProfileData is not the correct size");
-        u128 uuid{};
-        auto buffer = ctx.ReadBuffer();
-        std::memcpy(&uuid, buffer.data() + 8, sizeof(u128));
+    IPC::ResponseBuilder rb{ctx, 3};
+    rb.Push(ResultSuccess);
+    rb.PushRaw(*ipv4);
+}
+void IGeneralService::CreateTemporaryNetworkProfile(Kernel::HLERequestContext& ctx) {
```
Comment (liamwhite) [inline src/core/hle/service/nifm/nifm.cpp:384]: missing newline between function definitions

**Review by liamwhite — CHANGES_REQUESTED** [review]

File: src/dedicated_room/yuzu-room.cpp [hunk 23]
```diff
@@ -0,0 +1,375 @@
...
+#else
+        std::cout
+            << "yuzu Web Services is not available with this build: validation is disabled.\n\n";
+        verify_backend = std::make_unique<Network::VerifyUser::NullBackend>();
+#endif
+    } else {
+        verify_backend = std::make_unique<Network::VerifyUser::NullBackend>();
+    }
+
+    Core::System system{};
```
Comment (lioncash) [inline src/dedicated_room/yuzu-room.cpp:337]: It seems extremely suspect to need to construct the entire system state just to get the room network. Especially if that's all it's being used for

File: src/dedicated_room/yuzu-room.cpp [hunk 24]
```diff
@@ -0,0 +1,375 @@
...
+        std::cout << "preferred-game-id not set!\nThis should get set to allow users to find your "
+                     "room.\nSet with --preferred-game-id id\n\n";
+    }
+    if (max_members > Network::MaxConcurrentConnections || max_members < 2) {
+        std::cout << "max_members needs to be in the range 2 - "
+                  << Network::MaxConcurrentConnections << "!\n\n";
+        PrintHelp(argv[0]);
+        return -1;
+    }
+    if (port > 65535) {
```
Comment (lioncash) [inline src/dedicated_room/yuzu-room.cpp:278]:
```suggestion
    if (port > UINT16_MAX) {
```

File: src/dedicated_room/yuzu-room.cpp (hunk 2 above)
Comment (FearlessTobi) [reply src/dedicated_room/yuzu-room.cpp:1]: Since this code was ported with minimal changes from Citra, the copyright header needs to be preserved.

**Review by german77 — APPROVED** [review]

**Review by liamwhite — APPROVED** [review]

**Review by bunnei — APPROVED** [review]

**Comment by unuruu5760** [comment]
Local play still not working in luigi's mansion 3 scarescraper.

**Pull request merged**